README.md
# phpMyFAQ 3.2

![Packagist PHP Version Support](https://img.shields.io/packagist/php-v/thorsten/phpMyFAQ)
![GitHub](https://img.shields.io/github/license/thorsten/phpMyFAQ)
![GitHub commit activity](https://img.shields.io/github/commit-activity/m/thorsten/phpMyFAQ)

## What is phpMyFAQ?

phpMyFAQ is a multilingual, completely database-driven FAQ-system. It supports various databases to store all data, PHP
8+ is needed in order to access this data. phpMyFAQ also offers a multi-language Content Management System with a
WYSIWYG editor and a media manager, real time search support with Elasticsearch, flexible multi-user support with user
and group based permissions on categories and records, a wiki-like revision feature, a news system, user-tracking, 40+
supported languages, enhanced automatic content negotiation, HTML5/CSS3 based responsive templates, PDF-support, a
backup and restore system, a dynamic sitemap, related FAQs, tagging, enhanced SEO features, built-in spam protection
systems, OpenLDAP and Microsoft Active Directory support, and an easy-to-use installation and update script.

## Requirements

phpMyFAQ is only supported on PHP 8 and up, you need a database as well. Supported databases are MySQL, MariaDB,
Percona Server, PostgreSQL, Microsoft SQL Server and SQLite3. If you want to use Elasticsearch as main search
engine, you need Elasticsearch 5.x or later. Check our detailed requirements on
[phpmyfaq.de](https://www.phpmyfaq.de/requirements) for more information.

## Installation

### phpMyFAQ installation package for end-users

The best way to install phpMyFAQ is to download it on [phpmyfaq.de](https://www.phpmyfaq.de/download), unzip the package
and open http://www.example.org/phpmyfaq/setup/index.php in your preferred browser.

### phpMyFAQ installation with Docker

#### Dockerfile

The Dockerfile provided in this repo only build an environment to run any release it's for development purpose. It does
not contain any code as the phpmyfaq folder is meant to be mounted as the `/var/www/html` folder in the container.

To build a production release please use the [docker-hub](https://github.com/phpMyFAQ/docker-hub) repository or use
images provided on [docker.io](https://hub.docker.com/r/phpmyfaq/phpmyfaq/).

#### docker-compose.yml

For development purposes you can start a full stack to run your current PhpMyFAQ source code from your local repo.

    $ docker-compose up

The command above starts 10 containers for multi database development as following.

_Specific images started once to prepare the project:_

- **composer**: update composer dependencies
- **yarn**: update yarn dependencies

_Running using named volumes:_

- **mariadb**: image with MariaDB database with xtrabackup support
- **phpmyadmin**: a PHP tool to have a look on your MariaDB database.
- **postgres**: image with PostgreSQL database
- **pgadmin**: a PHP tool to have a look on your PostgreSQL database.
- **sqlserver**: image with Microsoft SQL Server for Linux
- **elasticsearch**: Open Source Software image (it means it does not have XPack installed)
- **openldap**: Open Source LDAP server
- **phpldapadmin**: a PHP tool to have a look on your LDAP database

_Running apache web server with PHP 8.0 support:_

- **phpmyfaq**: mounts the `phpmyfaq` folder in place of `/var/www/html`.

Then services will be available at following addresses:

- phpMyFAQ: (http://localhost:8080)
- phpMyAdmin: (http://localhost:8000)
- pgAdmin: (http://localhost:8008)
- phpLDAPadmin: (http://localhost:8088/)

#### Running tests

To run the test using Docker you have to install the Composer development dependencies

    $ curl -s https://getcomposer.org/installer | php
    $ php composer.phar install

#### Quote from ElasticSearch documentation

The vm.max_map_count kernel setting needs to be set to at least 262144 for production use. Depending on your platform:

##### Linux

The vm.max*map_count setting should be set permanently in */etc/sysctl.conf\_:

    $ grep vm.max_map_count /etc/sysctl.conf
    vm.max_map_count=262144

To apply the setting on a live system type: `sysctl -w vm.max_map_count=262144`

##### macOS with Docker for Mac

The vm.max_map_count setting must be set within the xhyve virtual machine:

    $ screen ~/Library/Containers/com.docker.docker/Data/com.docker.driver.amd64-linux/tty

Log in with root and no password. Then configure the sysctl setting as you would for Linux:

    $ sysctl -w vm.max_map_count=262144

##### Windows and macOS with Docker Toolbox

The vm.max_map_count setting must be set via docker-machine:

    $ docker-machine ssh
    $ sudo sysctl -w vm.max_map_count=262144

### phpMyFAQ local installation from Github

To run phpMyFAQ locally you need at least a running web server with PHP support and a database.

    $ git clone git://github.com/thorsten/phpMyFAQ.git
    $ cd phpMyFAQ
    $ curl -s https://getcomposer.org/installer | php
    $ php composer.phar install
    $ curl -o- -L https://yarnpkg.com/install.sh | bash
    $ yarn install
    $ yarn build

Then just open http://www.example.org/phpmyfaq/setup/index.php in your browser.

## Testing

To run our unit tests via PHPUnit v9.x, just execute this command on your CLI

    $ curl -s https://getcomposer.org/installer | php
    $ php composer.phar install
    $ ./vendor/bin/phpunit

Please note that phpMyFAQ needs to be installed via Composer.

## Versioning

For transparency and insight into our release cycle, and for striving to maintain backward compatibility, phpMyFAQ will
be maintained under the Semantic Versioning guidelines as much as possible.

Releases will be numbered with the following format:

`<major>.<minor>.<patch>`

And constructed with the following guidelines:

- Breaking backward compatibility bumps the major (and resets the minor and patch)
- New additions without breaking backward compatibility bumps the minor (and resets the patch)
- Bug fixes and misc changes bumps the patch

For more information on SemVer, please visit http://semver.org/.

## Bug tracker

Found a bug, or do you have a feature request? [Please open a new issue](https://github.com/thorsten/phpMyFAQ/issues).
Before opening any issue, please search for existing issues.

## Contributing

Please check out our page about contributing on [phpmyfaq.de](https://www.phpmyfaq.de/contribute).

## Documentation

You can find the full documentation on [phpmyfaq.de](https://www.phpmyfaq.de/documentation).

## REST API v2.1

The REST API v2.1 documentation is located [here in this repository](API.md) and on
[phpmyfaq.de](https://www.phpmyfaq.de/documentation).

## License

Mozilla Public License 2.0, see LICENSE for more information.

Copyright © 2001 - 2022 Thorsten Rinne and the phpMyFAQ Team


CHANGELOG
phpMyFAQ 3.0.11
Codename "Phobos"

CHANGELOG

This is a log of major user-visible changes in each phpMyFAQ release.

Version 3.0.12 - 2022-01-22
- fixed broken LDAP authentication (Thorsten)
- updated bundled dependencies (Thorsten)
- fixed minor bugs (Thorsten)

Version 3.0.11 - 2022-01-18
- fixed enabled debug mode (Thorsten)
- updated bundled dependencies (Thorsten)

Version 3.0.10 - 2022-01-17
- fixed multiple XSS and CSRF vulnerabilities (0x7zed, M0rphling, justinp09010, Dennis Yassine, Thorsten)
- fixed many minor bugs (Thorsten)

Version 3.0.9 - 2021-04-17
- fixed minor bugs (Thorsten)

Version 3.0.8 - 2021-02-24
- updated to Bootstrap v4.6 (Thorsten)
- fixed minor bugs (Thorsten)

Version 3.0.7 - 2020-12-23
- fixed XSS vulnerability (Curtis Robinson, Thorsten)
- added TOC plugin for TinyMCE (Thorsten)
- removed support for deprecated data-vocabulary.org schema (Thorsten)
- removed Travis CI build (Thorsten)
- fixed minor bugs (Thorsten)

Version 3.0.6 - 2020-11-27
- added support for PHP 8.0
- fixed minor bugs (Thorsten)

Version 3.0.5 - 2020-10-17
- minor improvements (Thorsten)
- fixed minor bugs (Thorsten)

Version 3.0.4 - 2020-07-26
- session timeout extended to 5 hours (Thorsten)
- fixed minor bugs (Thorsten)

Version 3.0.3 - 2020-05-21
- improved FAQ editing (Thorsten)
- updated to Bootstrap v4.5 (Thorsten)
- fixed minor bugs (Thorsten)

Version 3.0.2 - 2020-04-16
- improved handling of multiple homepage categories (Thorsten)
- improved FAQ editing (Thorsten)
- fixed minor bugs (Thorsten)

Version 3.0.1 - 2020-03-17
- re-added tag cloud on several pages (Thorsten)
- fixed minor bugs (Thorsten)
- fixed update issues for PostgreSQL (Thorsten)

Version 3.0.0 - 2020-02-16
- changed PHP requirement to PHP 7.2+ (Thorsten)
- added PHP namespaces (Thorsten)
- added Docker support (Adrien Estanove)
- added experimental large permissions support with sections (Timo Wolf, Thorsten)
- added support for Elasticsearch v5+ (Thorsten)
- added LDAP configuration frontend (Thorsten)
- added configuration for enable/disable XML sitemap (Thorsten)
- added support for category images (Thorsten)
- added support for categories on homepage (Thorsten)
- added filter functionality in templates (Thorsten)
- added improved attachment overview (Thorsten)
- added HTML5 export (Thorsten)
- added support for EU General Data Protection Regulation (Jochen Steinhilber)
- added multiple attachment upload (Thorsten)
- added 404 page template (Thorsten)
- added Mongolian translation (khaidaw@gmail.com)
- added support for adding own meta content in templates (Thorsten)
- added new REST API v2 (includes login) (Thorsten)
- improved sticky records (Thorsten)
- improved brute force handling (Thorsten)
- switched CSS development from LESS to SCSS (Thorsten)
- template variable syntax compatible with Twig/Handlebars (Thorsten)
- updated Turkish translation (Can Kirca)
- updated bundled dependencies (Thorsten)
- deprecated RSS feeds (Thorsten)
- deprecated REST API v1 (Thorsten)
- removed translation admin frontend (Thorsten)
- removed XHTML export (Thorsten)
- removed support for ext/mssql (Thorsten)
- removed bundled Symfony ClassLoader (Thorsten)
- removed Bower, now using Yarn only (Thorsten)
- removed Grunt, now using Webpack (Thorsten)
- removed Modernizr (Thorsten)
- removed share on Facebook link (Thorsten)
- removed Facebook Like Button support (Thorsten)

Version 2.9.13 - 2019-02-14
- fixed XSS vulnerabilities in the bundled Bootstrap version (Thorsten)

Version 2.9.12 - 2019-02-12
- updated bundled Bootstrap to version 3.4.0 (Thorsten)
- fixed minor bugs (Thorsten)

Version 2.9.11 - 2018-09-02
- fixed multiple vulnerabilities (Thorsten)
- updated bundled jQuery to version 1.12.4 (Thorsten)
- updated bundled Handlebars to version 4.0.11 (Thorsten)

Version 2.9.10 - 2018-02-17
- updated Dutch translation (https://github.com/joskevos)
- updated bundled SwiftMailer to version 5.4.9 (Thorsten)
- updated bundled phpseclib to version 2.0.9 (Thorsten)
- updated bundled TinyMCE to version 4.6.7 (Thorsten)
- fixed minor bugs (Thorsten)

Version 2.9.9 - 2017-10-19
- fixed multiple XSS and CSRF vulnerabilities (Thorsten)
- updated bundled Bootstrap to version 3.3.7 (Thorsten)
- updated bundled Font Awesome to version 4.7.0 (Thorsten)
- updated bundled TinyMCE to version 4.5.7 (Thorsten)
- updated bundled HighlightJS to version 9.12.0 (Thorsten)
- fixed minor bugs (Thorsten)

Version 2.9.8 - 2017-07-12
- fixed improper restriction (Thorsten)
- add LDAP search in sub groups (Thorsten)
- updated French translation
- fixed minor bugs (Thorsten)

Version 2.9.7 - 2017-04-02
- fixed stored XSS vulnerability (Thorsten)
- fixed minor bugs (Thorsten)

Version 2.9.6 - 2017-01-27
- fixed possible arbitrary PHP code execution (Thorsten)
- ready for PHP 7.1 (Thorsten)
- fixed minor bugs (Thorsten)

Version 2.9.5 - 2016-08-31
- fixed minor bugs (Thorsten, Matt Brennan)

Version 2.9.4 - 2016-08-02
- deactivated debug mode (Thorsten)
- updated bundled TinyMCE to version 4.4.1 (Thorsten)
- updated bundled Bootstrap to version 3.3.7 (Thorsten)
- fixed minor bugs (Thorsten)

Version 2.9.3 - 2016-08-01
- fixed minor bugs (Thorsten)

Version 2.9.2 - 2016-07-05
- updated bundled phpseclib to version 2.0.2 (Thorsten)
- updated bundled TinyMCE to version 4.4.0 (Thorsten)
- fixed minor bugs (Thorsten)

Version 2.9.1 - 2016-05-31
- fixed stored XSS vulnerability (Kacper Szurek)
- added new source code paste plugin (Thorsten)
- removed American English translation (Thorsten)
- updated bundled TinyMCE to version 4.3.12 (Thorsten)
- updated bundled Typeahead.js to version 0.11.0 (Thorsten)
- fixed minor bugs (Thorsten)

Version 2.9.0 - 2016-05-13
- ready for PHP 7 (Thorsten)
- changed PHP requirement to PHP 5.5+ and PHP 7 (Thorsten)
- added support for HHVM 3.4.2+ (Thorsten)
- added support for Elasticsearch (Thorsten)
- added new mobile first, touch-friendly default layout (Thorsten)

phpmyfaq/api.php
<?php

/**
 * The REST API
 *
 * This Source Code Form is subject to the terms of the Mozilla Public License,
 * v. 2.0. If a copy of the MPL was not distributed with this file, You can
 * obtain one at https://mozilla.org/MPL/2.0/.
 *
 * @package phpMyFAQ
 * @author Thorsten Rinne <thorsten@phpmyfaq.de>
 * @copyright 2009-2022 phpMyFAQ Team
 * @license http://www.mozilla.org/MPL/2.0/ Mozilla Public License Version 2.0
 * @link https://www.phpmyfaq.de
 * @since 2009-09-03
 */

const IS_VALID_PHPMYFAQ = null;

use phpMyFAQ\Attachment\AttachmentException;
use phpMyFAQ\Attachment\AttachmentFactory;
use phpMyFAQ\Category;
use phpMyFAQ\Category\CategoryRelation;
use phpMyFAQ\Comments;
use phpMyFAQ\Entity\CommentType;
use phpMyFAQ\Entity\FaqEntity;
use phpMyFAQ\Faq;
use phpMyFAQ\Faq\FaqMetaData;
use phpMyFAQ\Faq\FaqPermission;
use phpMyFAQ\Filter;
use phpMyFAQ\Helper\HttpHelper;
use phpMyFAQ\Helper\RegistrationHelper;
use phpMyFAQ\Language;
use phpMyFAQ\Language\Plurals;
use phpMyFAQ\News;
use phpMyFAQ\Permission\MediumPermission;
use phpMyFAQ\Question;
use phpMyFAQ\Search;
use phpMyFAQ\Search\SearchResultSet;
use phpMyFAQ\Services;
use phpMyFAQ\Strings;
use phpMyFAQ\Tags;
use phpMyFAQ\User\CurrentUser;
use phpMyFAQ\Utils;

//
// Bootstrapping
//
require 'src/Bootstrap.php';

//
// Send headers
//
$http = new HttpHelper();
$http->setContentType('application/json');
$http->fetchAllHeaders();
$http->addHeader();
//
// Set user permissions
//
$auth = false;

$action = Filter::filterInput(INPUT_GET, 'action', FILTER_UNSAFE_RAW);
$lang = Filter::filterInput(INPUT_GET, 'lang', FILTER_UNSAFE_RAW, 'en');
$categoryId = Filter::filterInput(INPUT_GET, 'categoryId', FILTER_VALIDATE_INT);
$recordId = Filter::filterInput(INPUT_GET, 'recordId', FILTER_VALIDATE_INT);
$tagId = Filter::filterInput(INPUT_GET, 'tagId', FILTER_VALIDATE_INT);

$faqUsername = Filter::filterInput(INPUT_POST, 'username', FILTER_UNSAFE_RAW);
$faqPassword = Filter::filterInput(INPUT_POST, 'password', FILTER_UNSAFE_RAW, FILTER_FLAG_NO_ENCODE_QUOTES);

//
// Get language (default: english)
//
$language = new Language($faqConfig);
$currentLanguage = $language->setLanguageByAcceptLanguage();

//
// Set language
//
if (Language::isASupportedLanguage($currentLanguage)) {
    require PMF_LANGUAGE_DIR . '/language_' . $currentLanguage . '.php';
} else {
    require PMF_LANGUAGE_DIR . '/language_en.php';
}
$faqConfig->setLanguage($language);

$plr = new Plurals($PMF_LANG);
Strings::init($currentLanguage);

//
// Set empty result
$result = [];

//
// Check if user is already authenticated
//
if (is_null($faqUsername) && is_null($faqPassword)) {
    $user = CurrentUser::getFromCookie($faqConfig);
    if (!$user instanceof CurrentUser) {
        $user = CurrentUser::getFromSession($faqConfig);
    }
    if ($user instanceof CurrentUser) {
        $auth = true;
    } else {
        $user = new CurrentUser($faqConfig);
    }
} else {
    $user = new CurrentUser($faqConfig);
}

//
// Get current user and group id - default: -1
//
if (!is_null($user) && $user instanceof CurrentUser) {
    $currentUser = $user->getUserId();
    if ($user->perm instanceof MediumPermission) {
        $currentGroups = $user->perm->getUserGroups($currentUser);
    } else {
        $currentGroups = [-1];
    }
    if (0 == count($currentGroups)) {
        $currentGroups = [-1];
    }
} else {
    $currentUser = -1;
    $currentGroups = [-1];
}

//
// Handle actions
//
switch ($action) {
    //
    // v2.1
    //
    case 'version':
        $result = $faqConfig->getVersion();
        break;

    case 'language':
        $result = $faqConfig->getLanguage()->getLanguage();
        break;

    case 'search':
        $user = new CurrentUser($faqConfig);
        $search = new Search($faqConfig);
        $search->setCategory(new Category($faqConfig));

        $faqPermission = new FaqPermission($faqConfig);
        $faqSearchResult = new SearchResultSet($user, $faqPermission, $faqConfig);

        $searchString = Filter::filterInput(INPUT_GET, 'q', FILTER_UNSAFE_RAW);
        $searchResults = $search->search($searchString, false);
        $faqSearchResult->reviewResultSet($searchResults);
        if ($faqSearchResult->getNumberOfResults() > 0) {
            $url = $faqConfig->getDefaultUrl() . 'index.php?action=faq&cat=%d&id=%d&artlang=%s';
            foreach ($faqSearchResult->getResultSet() as $data) {
                $data->answer = html_entity_decode(strip_tags($data->answer), ENT_COMPAT, 'utf-8');
                $data->answer = Utils::makeShorterText($data->answer, 12);
                $data->link = sprintf($url, $data->category_id, $data->id, $data->lang);
                $result[] = $data;
            }
        } else {
            $http->setStatus(404);
        }

        break;

    case 'categories':
        $category = new Category($faqConfig, $currentGroups, true);
        $category->setUser($currentUser);
        $category->setGroups($currentGroups);
        $category->setLanguage($currentLanguage);
        $result = array_values($category->getAllCategories());
        if (count($result) === 0) {
            $http->setStatus(404);
        }
        break;

    case 'tags':
        $tags = new Tags($faqConfig);
        $result = $tags->getPopularTagsAsArray(16);
        if (count($result) === 0) {
            $http->setStatus(404);
        }
        break;

    case 'open-questions':
        $questions = new Question($faqConfig);
        $result = $questions->getAllOpenQuestions();
        if (count($result) === 0) {
            $http->setStatus(404);
        }
        break;

    case 'searches':
        $search = new Search($faqConfig);
        $result = $search->getMostPopularSearches(7, true);
        if (count($result) === 0) {
            $http->setStatus(404);
        }
        break;

    case 'comments':
        $comment = new Comments($faqConfig);
        $result = $comment->getCommentsData($recordId, CommentType::FAQ);
        if (count($result) === 0) {
            $http->setStatus(404);
        }
        break;

    case 'attachments':
        $attachments = $result = [];
        try {
            $attachments = AttachmentFactory::fetchByRecordId($faqConfig, $recordId);
        } catch (AttachmentException $e) {
            $result = [];
        }
        foreach ($attachments as $attachment) {
            $result[] = [
                'filename' => $attachment->getFilename(),
                'url' => $faqConfig->getDefaultUrl() . $attachment->buildUrl(),
            ];
        }
        if (count($result) === 0) {
            $http->setStatus(404);
        }
        break;

    case 'news':
        $news = new News($faqConfig);
        $result = $news->getLatestData(false, true, true);
        if (count($result) === 0) {
            $http->setStatus(404);
        }
        break;


    case 'faqs':
        $filter = Filter::filterInput(INPUT_GET, 'filter', FILTER_UNSAFE_RAW);
        $faq = new Faq($faqConfig);
        $faq->setUser($currentUser);
        $faq->setGroups($currentGroups);

        // api/v2.1/faqs/:categoryId
        if (!is_null($categoryId)) {
            try {
                $result = $faq->getAllRecordsByCategoryId($categoryId);
            } catch (Exception $e) {
                $http->setStatus(400);
            }
        }

        // api/v2.1/faqs/tags/:tagId
        if (!is_null($tagId)) {
            $tags = new Tags($faqConfig);
            $recordIds = $tags->getFaqsByTagId($tagId);
            try {
                $result = $faq->getRecordsByIds($recordIds);
            } catch (Exception $e) {
                $http->setStatus(400);
            }
        }

        // api/v2.1/faqs/popular
        if ('popular' === $filter) {
            $result = array_values($faq->getTopTenData(PMF_NUMBER_RECORDS_TOPTEN));
        }

        // api/v2.1/faqs/latest
        if ('latest' === $filter) {
            $result = array_values($faq->getLatestData(PMF_NUMBER_RECORDS_LATEST));
        }

        // api/v2.1/faqs/sticky
        if ('sticky' === $filter) {
            $result = array_values($faq->getStickyRecordsData());
        }

        // api/v2.1/faqs
        if (is_null($categoryId) && is_null($tagId) && is_null($filter)) {
            $faq->getAllRecords(FAQ_SORTING_TYPE_CATID_FAQID, ['lang' => $currentLanguage]);
            $result = $faq->faqRecords;
        }

        if (count($result) === 0) {
            $http->setStatus(404);
        }
        break;


    case 'faq':
        //
        // GET
        //
        $filter = Filter::filterInput(INPUT_GET, 'filter', FILTER_UNSAFE_RAW);
        $faq = new Faq($faqConfig);
        $faq->setUser($currentUser);
        $faq->setGroups($currentGroups);

        if ($recordId > 0) {
            $faq->getRecord($recordId);
            $result = $faq->faqRecord;

            if (count($result) === 0 || $result['solution_id'] === 42) {
                $result = new stdClass();
                $http->setStatus(404);
            }

            if ('pdf' === $filter) {
                $service = new Services($faqConfig);
                $service->setFaqId($recordId);
                $service->setLanguage($currentLanguage);
                $service->setCategoryId($categoryId);

                $result = $service->getPdfApiLink();
            }
        }

        //
        // POST
        //
        if ($faqConfig->get('api.apiClientToken') !== $http->getClientApiToken()) {
            $http->setStatus(401);
            $result = [
                'stored' => false,
                'error' => 'X_PMF_Token not valid.'
            ];
        }

        $languageCode = Filter::filterInput(INPUT_POST, 'language', FILTER_UNSAFE_RAW);
        $categoryId = Filter::filterInput(INPUT_POST, 'category-id', FILTER_VALIDATE_INT);
        $question = Filter::filterInput(INPUT_POST, 'question', FILTER_UNSAFE_RAW);
        $answer = Filter::filterInput(INPUT_POST, 'answer', FILTER_UNSAFE_RAW);
        $keywords = Filter::filterInput(INPUT_POST, 'keywords', FILTER_UNSAFE_RAW);
        $author = Filter::filterInput(INPUT_POST, 'author', FILTER_UNSAFE_RAW);
        $email = Filter::filterInput(INPUT_POST, 'email', FILTER_UNSAFE_RAW);
        $isActive = Filter::filterInput(INPUT_POST, 'is-active', FILTER_UNSAFE_RAW);
        $isSticky = Filter::filterInput(INPUT_POST, 'is-sticky', FILTER_UNSAFE_RAW);

        $categories = [ $categoryId ];
        $isActive = $isActive === 'true';
        $isSticky = $isSticky === 'true';

        $faqData = new FaqEntity();
        $faqData
            ->setLanguage($languageCode)
            ->setQuestion($question)
            ->setAnswer($answer)
            ->setKeywords($keywords)
            ->setAuthor($author)
            ->setEmail($email)
            ->setActive($isActive)
            ->setSticky($isSticky)
            ->setComment(false)
            ->setLinkState('')
            ->setNotes('');

        $faqId = $faq->create($faqData);

        $faqMetaData = new FaqMetaData($faqConfig);
        $faqMetaData
            ->setFaqId($faqId)
            ->setFaqLanguage($languageCode)
            ->setCategories($categories)
            ->save();

        $result = [
            'stored' => true
        ];
        break;

    case 'login':
        $currentUser = new CurrentUser($faqConfig);

        if ($currentUser->login($faqUsername, $faqPassword)) {
            if ($currentUser->getStatus() !== 'blocked') {
                $auth = true;
                $result = [
                    'loggedin' => true
                ];
            } else {
                $auth = false;
                $http->setStatus(400);
                $result = [
                    'loggedin' => false,
                    'error' => $PMF_LANG['ad_auth_fail']
                ];
            }
        } else {
            $auth = false;
            $http->setStatus(400);
            $result = [
                'loggedin' => false,
                'error' => $PMF_LANG['ad_auth_fail']
            ];
        }
        break;

    case 'register':
        if ($faqConfig->get('api.apiClientToken') !== $http->getClientApiToken()) {
            $http->setStatus(401);
            $result = [
                'registered' => false,
                'error' => 'X_PMF_Token not valid.'
            ];
            break;
        }

        $registration = new RegistrationHelper($faqConfig);

        $userName = Filter::filterInput(INPUT_POST, 'username', FILTER_UNSAFE_RAW);
        $fullName = Filter::filterInput(INPUT_POST, 'fullname', FILTER_UNSAFE_RAW);
        $email = Filter::filterInput(INPUT_POST, 'email', FILTER_SANITIZE_EMAIL);
        $isVisible = Filter::filterInput(INPUT_POST, 'is-visible', FILTER_UNSAFE_RAW);
        $isVisible = $isVisible === 'true';

        if (!$registration->isDomainWhitelisted($email)) {
            $http->setStatus(400);
            $result = [
                'registered' => false,
                'error' => 'The domain is not whitelisted.'
            ];
            break;
        }

        if (!is_null($userName) && !is_null($fullName) && !is_null($email)) {
            $result = $registration->createUser($userName, $fullName, $email, $isVisible);
            $http->setStatus(200);
        } else {
            $http->setStatus(400);
            $result = [
                'registered' => false,
                'error' => $PMF_LANG['err_sendMail']
            ];
        }

        break;
}

//
// Check if FAQ should be secured
//
if (!$auth && $faqConfig->get('security.enableLoginOnly')) {
    $http->setStatus(403);
    $http->sendJsonWithHeaders([ 'error' => 'You are not allowed to view this content.' ]);
    exit();
}

$http->sendJsonWithHeaders($result);
